File: 백준/Silver/11725. 트리의 부모 찾기/트리의 부모 찾기.py
import sys
sys.setrecursionlimit(10**8)
input = sys.stdin.readline



def dfs(n):
    for i in graph[n]:
        if v[i] == 0:
            v[i] = n
            dfs(i)





N = int(input())
graph = [[] for _ in range(N+1)]
v = [0]*(N+1)

for _ in range(N-1):
    x,y = list(map(int, input().strip().split()))
    graph[x].append(y)
    graph[y].append(x)


dfs(1)


for i in range(2, N+1):
    print(v[i])

File: 백준/Silver/11725. 트리의 부모 찾기/트리의 부모 찾기.java
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.BufferedReader;

import java.util.ArrayList;
import java.util.StringTokenizer;

public class Main {
    static ArrayList<Integer>[] graph;
    static int[] parents;
    static boolean[] v;


    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int N = Integer.parseInt(br.readLine());

        // graph
        graph = new ArrayList[N+1];
        for (int i = 1; i < N+1; i++) {
            graph[i] = new ArrayList<>();
        }

        // link
        for (int i = 0; i < N-1; i++) {
            StringTokenizer st = new StringTokenizer(br.readLine(), " ");
            int x = Integer.parseInt(st.nextToken());
            int y = Integer.parseInt(st.nextToken());
            graph[x].add(y);
            graph[y].add(x);
        }

        parents = new int[N+1];
        v = new boolean[N+1];

        // start
        dfs(1);

        // print parents
        StringBuilder sb = new StringBuilder();
        for (int i = 2; i < N+1; i++) {
            sb.append(parents[i]).append('\n');
        }
        System.out.print(sb);

        
    }





    // DFS
    static void dfs(int i) {
        v[i] = true;
        for (int n : graph[i]) {
            if (!v[n]) {
                parents[n] = i;
                dfs(n);
            }
        }
    }





}